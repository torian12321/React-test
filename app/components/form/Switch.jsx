var React = require('react');

class Radio extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            checked   : this.props.checked ? 'checked' : null,
        };
    }
    click(){
        this.setState({
            checked: this.state.checked ? null : 'checked'
        });
    }
    render() {
        return (
            <div className={ 'form-switch ' + this.props.style }>
                <input
                    type   = "checkbox"
                    checked= { this.state.checked }
                />
                <label onClick= { this.click.bind(this) }></label>
            </div>
        );
    }
}


Radio.propTypes = {
    checked : React.PropTypes.bool.isRequired,
    style   : React.PropTypes.string.isRequired,
};
Radio.defaultProps = {
    checked: false,
    style  : ''       // texted
};

module.exports = Radio;
