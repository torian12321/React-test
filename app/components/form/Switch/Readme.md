Basic
```
<div>
	<Switch />
	<Switch checked={false}/>
	<Switch checked={true}/>
</div>
```

Texted
```
<Switch texted={true}/>
```