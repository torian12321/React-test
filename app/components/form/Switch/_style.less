@import '~LESS_conf';

.form-switch{
    @BG_1           : @colorState_disable;
    @BG_2           : @themeColor_2;
	@height         : @inputHeight;
	@transition-time: 0.20s;

    font-size  : @height;
    line-height: 1;
	input[type="checkbox"]{
		opacity : 0;
		position: absolute;
		& + label{
			position  : relative;
			display   : block;
			width     : 2em;
			height    : 1em;
			background: @BG_1;
			cursor    : pointer;
			.rounded(1em);
			.transition(@property: background-color; @time: @transition-time);
			&:after{
				position  : absolute;
				content   : '';
				left      : .15em;
				top       : .15em;
				height    : .7em;
				width     : .7em;
				background: darken(@BG_1, 20%);
				.rounded(50%);
				.transition(@time: @transition-time);
			}
		}

		//----------------------
		//----    STATES    ----
		//----------------------

		// Checked
		&:checked + label{
			&:after{
				left: 1.15em;
				background: @themeColor_1;
			}
		}

		// Focus
	    // Hover
		&:focus + label:after,
		&:hover + label:after{
			background: @BG_2;
		}

		// Disabled
		&:disabled + label{
			color : @fontColor_disable;
        	cursor: not-allowed;
			&:after{
				background: darken(@colorState_disable, 30);
			}
		}	
	}

	&.texted input[type="checkbox"]{
		& + label{
			&:before{
				position   : absolute;
				content    : 'Off';
				right      : 12.5%;
				font-size  : .5em;
				line-height: 2em;
                .smart-text-color(@BG_1);
                .transition(@property: right; @time: @transition-time);
			}
		}
		&:checked + label{
			&:before{
				content: 'On';
				right  : 55%;
			}
		}
	}
}