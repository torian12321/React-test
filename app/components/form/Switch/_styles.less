@import '~LESS_conf';

//TODO 
//-> change margins+paddings+fonts to em
//-> instead of display:none, try opacity:0+ positioon absolute
//        That will do the switch selectable by keyboard

// .form-switch{
// 	@height         : @inputHeight;
// 	@width          : @height * 2;
// 	@ball-height    : @height * 0.8;
// 	@transition-time: 0.20s;

// 	input[type="checkbox"]{
// 		display: none;

// 		& + label{
// 			position  : relative;
// 			display   : block;
// 			width     : @width;
// 			height    : @height;
// 			background: @colorState_disable;
// 			cursor    : pointer;
// 			.rounded(@height);
// 			.transition(@property: background-color; @time: @transition-time);
// 			&:after{
// 				position  : absolute;
// 				content   : '';
// 				left      : (@height - @ball-height) / 2;
// 				top       : (@height - @ball-height) / 2;
// 				height    : @ball-height;
// 				width     : @ball-height;
// 				background: @themeColor_1;
// 				.rounded(50%);
// 				.transition(@transition-time);
// 			}
// 		}
// 		&:checked + label{
// 			background: @themeColor_2;
// 			&:after{
// 				left: @width - @ball-height - ((@height - @ball-height) / 2);
// 			}
// 		}
// 	}

// 	&.texted input[type="checkbox"]{
// 		& + label{
// 			&:before{
// 				position   : absolute;
// 				content    : 'Off';
// 				right      : 12.5%;
// 				font-size  : @height * 0.5;
// 				line-height: @height;
// 				color      : @white;
// 				.transition(@transition-time);
// 			}
// 		}
// 		&:checked + label{
// 			&:before{
// 				content: 'On';
// 				right  : 55%;
// 			}
// 		}
// 	}
// }

.form-switch{
    @BG_1           : @colorState_disable;
    @BG_2           : @themeColor_2;
	@height         : @inputHeight;
	@transition-time: 0.20s;

    font-size  : @height;
    line-height: 1;
	input[type="checkbox"]{
		display: none;

		& + label{
			position  : relative;
			display   : block;
			width     : 2em;
			height    : 1em;
			background: @BG_1;
			cursor    : pointer;
			.rounded(1em);
			.transition(@property: background-color; @time: @transition-time);
			&:after{
				position  : absolute;
				content   : '';
				left      : .1em;
				top       : .1em;
				height    : .8em;
				width     : .8em;
				background: @themeColor_1;
				.rounded(50%);
				.transition(@property: left; @time: @transition-time);
			}
		}
		&:checked + label{
			background: @BG_2;
			&:after{
				left: 1.1em;
			}
		}
	}

	&.texted input[type="checkbox"]{
		& + label{
			&:before{
				position : absolute;
				content  : 'Off';
				right    : 12.5%;
				font-size: .5em;
                .smart-text-color(@BG_1);
				//.transition(@transition-time);
                .transition(@property: right; @time: @transition-time);
			}
		}
		&:checked + label{
			&:before{
				content: 'On';
				right  : 55%;
                .smart-text-color(@BG_2);
			}
		}
	}
}