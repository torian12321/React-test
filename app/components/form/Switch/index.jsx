var
React     = require('react'),
PropTypes = require('prop-types'),
classNames= require('classnames'),
uniqid    = require('uniqid');

class Switch extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            checked: this.props.checked,
        };
    }
    componentWillReceiveProps(nextProps) {
        if (nextProps.checked !== this.state.checked) {
            this.setState({ checked: nextProps.checked });
        }
    }

    handleClick(){
        this.setState({
            checked: !this.state.checked
        });
    }
    render() {
        const id = this.props.id || uniqid('switch_');

        return (
            <div className={classNames('form-switch', this.props.style)}>
                <input
                    id     = {id}
                    type   = "checkbox"
                    checked= {this.state.checked ? 'checked' : null}
                />
                <label htmlFor={id} />
            </div>
        );
    }
}


Switch.propTypes = {
    id     : PropTypes.string,
    checked: PropTypes.bool.isRequired,
    style  : PropTypes.oneOf(['texted'])
};
Switch.defaultProps = {
    checked: false
};

module.exports = Switch;