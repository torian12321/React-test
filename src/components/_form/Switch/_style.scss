@import '../../../styles/conf/index.scss';

// Vars
$formSwitch-color-bg       : $colorState_disable !default;
$formSwitch-color-main     : $themeColor_2 !default;
$formSwitch-color-sec      : $themeColor_1 !default;
$formSwitch-height         : $input-height !default;
$formSwitch-transition-time: .2s !default;

// Block
.form-switch {
  font-size  : $formSwitch-height;
  line-height: 1;

  &__cb {
    opacity : 0;
    position: absolute;
  }
  &__ball {
    position  : relative;
    display   : block;
    width     : 2em;
    height    : 1em;
    background: $formSwitch-color-bg;
    cursor    : pointer;
    @include rounded(1em);
    @include transition($property: background-color, $time: $formSwitch-transition-time);
    
    &:after{
      position  : absolute;
      content   : '';
      left      : .15em;
      top       : .15em;
      height    : .7em;
      width     : .7em;
      background: darken($formSwitch-color-bg, 20%);
      @include rounded-circular();
      @include transition($time: $formSwitch-transition-time);
    }
  }

  //----  Modifiers  -----

  // Active
  &--checked .form-switch__ball:after {
    left      : 1.15em;
    background: $formSwitch-color-sec;
  }

  // Focus
  // Hover
  &__cb:focus,
  &__cb:hover {
    + .form-switch__ball:after {
      background: $formSwitch-color-main;
    }
  }
  
  // Disabled
  &--disabled .form-switch__ball {
    color : $fontColor_disable;
    cursor: not-allowed;
    
    &:after {
      background: darken($colorState_disable, 30) !important;
    }
  }	
	

  // Texted version
	&--texted {
		.form-switch__ball:before {
      position   : absolute;
      content    : 'Off';
      right      : 12.5%;
      font-size  : .5em;
      line-height: 2em;
      @include smart-text-color($formSwitch-color-bg);
      @include transition($property: right, $time: $formSwitch-transition-time);
		}
		&.form-switch--checked .form-switch__ball:before {
      content: 'On';
      right  : 55%;
		}
	}
}