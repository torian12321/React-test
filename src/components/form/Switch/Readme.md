#### Basic
```
<div>
	<Switch />
	<Switch checked />
	<Switch checked={false} />
</div>
```

#### Texted
```
<Switch texted />
```

On change events
```
<Switch onChange={(v) => alert('The checkbox value is: ' + v)} />
```

#### Disabled
```
<div>
	<Switch disabled />
	<Switch checked disabled />
</div>
```