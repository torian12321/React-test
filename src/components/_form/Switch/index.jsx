import React      from 'react';
import PropTypes  from 'prop-types';
import classNames from 'classnames';
import uniqid     from 'uniqid';
import './_style.scss';

class Switch extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: this.props.checked,
    };

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(){
    this.setState({
      checked : !this.state.checked
    }, () => this.props.onChange(this.state.checked));
  }
  render() {
    const id = this.props.id || uniqid('switch_');

    return (
      <div className={classNames(
        'form-switch',
        this.state.checked ? 'form-switch--checked' : null,
        this.props.texted  ? 'form-switch--texted'  : null,
        this.props.disabled? 'form-switch--disabled': null,
        this.props.className
      )}>
        <input
          id       = {id}
          className= 'form-switch__cb'
          type     = "checkbox"
          checked  = {this.state.checked}
          disabled = {this.props.disabled}
          onChange = {this.handleChange}
        />
        <label htmlFor={id} className= 'form-switch__ball' />
      </div>
    );
  }
}


Switch.propTypes = {
  id       : PropTypes.string,
  className: PropTypes.string,
  checked  : PropTypes.bool,
  texted   : PropTypes.bool,
  disabled : PropTypes.bool,
  onChange : PropTypes.func
};
Switch.defaultProps = {
  checked : false,
  texted  : false,
  disabled: false,
  onChange: () => null
};

module.exports = Switch;