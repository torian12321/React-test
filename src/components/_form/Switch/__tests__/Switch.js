import React    from 'react';
import renderer from 'react-test-renderer';
import Switch   from '../index';

describe('component', () => {
  describe('<Switch />', () => {

    it('should render correctly', () => {
      expect(renderer.create(
        <Switch id='sw' />
      )).toMatchSnapshot();
    });

    it('should trigger onChange', () => {
      const changeHander = jest.fn();
      const comp = renderer.create(
        <Switch id='sw' onChange={changeHander} />
      ).toJSON();

      comp.children[0].props.onChange();
      expect(changeHander).toBeCalled();
      expect(comp).toMatchSnapshot();
    });



    it('should render checked', () => {
      expect(renderer.create(
        <Switch id='sw' checked />
      )).toMatchSnapshot();
    });

    it('should render texted', () => {
      expect(renderer.create(
        <Switch id='sw' texted />
      )).toMatchSnapshot();
    });

    it('should render disabled', () => {
      expect(renderer.create(
        <Switch id='sw' disabled />
      )).toMatchSnapshot();
    });

  });
});