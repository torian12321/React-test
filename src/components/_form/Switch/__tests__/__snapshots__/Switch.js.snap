// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <Switch /> should render checked 1`] = `
<div
  className="form-switch form-switch--checked"
>
  <input
    checked={true}
    className="form-switch__cb"
    disabled={false}
    id="sw"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="form-switch__ball"
    htmlFor="sw"
  />
</div>
`;

exports[`component <Switch /> should render correctly 1`] = `
<div
  className="form-switch"
>
  <input
    checked={false}
    className="form-switch__cb"
    disabled={false}
    id="sw"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="form-switch__ball"
    htmlFor="sw"
  />
</div>
`;

exports[`component <Switch /> should render disabled 1`] = `
<div
  className="form-switch form-switch--disabled"
>
  <input
    checked={false}
    className="form-switch__cb"
    disabled={true}
    id="sw"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="form-switch__ball"
    htmlFor="sw"
  />
</div>
`;

exports[`component <Switch /> should render texted 1`] = `
<div
  className="form-switch form-switch--texted"
>
  <input
    checked={false}
    className="form-switch__cb"
    disabled={false}
    id="sw"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="form-switch__ball"
    htmlFor="sw"
  />
</div>
`;

exports[`component <Switch /> should trigger onChange 1`] = `
<div
  className="form-switch"
>
  <input
    checked={false}
    className="form-switch__cb"
    disabled={false}
    id="sw"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="form-switch__ball"
    htmlFor="sw"
  />
</div>
`;
